server/src/main/java/org/opensearch/node/PerformanceCollectorService.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.node;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.ClusterChangedEvent;
+import org.opensearch.cluster.ClusterStateListener;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.lifecycle.AbstractLifecycleComponent;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.util.concurrent.ConcurrentCollections;
+import org.opensearch.ratelimiting.tracker.NodePerformanceTracker;
+import org.opensearch.threadpool.Scheduler;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * This collects node level performance statistics such as cpu, memory, IO of each node and makes it available for
+ * coordinator node to aid in throttling, ranking etc
+ */
+public class PerformanceCollectorService extends AbstractLifecycleComponent implements ClusterStateListener {
+
+    /**
+     * This refresh interval denotes the polling interval of PerformanceCollectorService to refresh the performance stats
+     * from local node
+     */
+    private static long REFRESH_INTERVAL_IN_MILLIS = 1000;
+
+    private static final Logger logger = LogManager.getLogger(PerformanceCollectorService.class);
+    private final ConcurrentMap<String, NodePerformanceStats> nodeIdToPerfStats = ConcurrentCollections.newConcurrentMap();
+
+    private ThreadPool threadPool;
+    private volatile Scheduler.Cancellable scheduledFuture;
+
+    private NodePerformanceTracker nodePerformanceTracker;
+    private ClusterService clusterService;
+
+    public PerformanceCollectorService(
+        NodePerformanceTracker nodePerformanceTracker,
+        ClusterService clusterService,
+        ThreadPool threadPool
+    ) {
+        this.threadPool = threadPool;
+        this.nodePerformanceTracker = nodePerformanceTracker;
+        this.clusterService = clusterService;
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.nodesRemoved()) {
+            for (DiscoveryNode removedNode : event.nodesDelta().removedNodes()) {
+                removeNodePerfStatistics(removedNode.getId());
+            }
+        }
+    }
+
+    void removeNodePerfStatistics(String nodeId) {
+        nodeIdToPerfStats.remove(nodeId);
+    }
+
+    /**
+     * Collect node performance statistics along with the timestamp
+     */
+    public void collectNodePerfStatistics(String nodeId, long timestamp, double memoryUtilizationPercent, double cpuUtilizationPercent) {
```

Comment:
Do you think we can extend he behaviour from ResponseCollectorService that does similar ranking calculation?

Reply:
Yeah that's the plan but we didn't include it as part of this PR. As part of POC , we pass the rank we get from 'ResponseCollectorService' and we modified the rank in this collector based on performance stats.

In this comment, do you mean to extract interface / base class (or) extending response collector service here ?